File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-5.java
class Solution {
    public int lengthOfLongestSubstring(String s) {

        HashMap<Character, Integer> mapping = new HashMap<>();
        int longest = 0;
        int curr = 0;
        int left = 0;

        for (int right = 0; right < s.length(); right++){
            if (mapping.containsKey(s.charAt(right))){
                mapping.put(s.charAt(right), mapping.get(s.charAt(right))+1);
            }
            else {
                mapping.put(s.charAt(right), 1);
            }

            while (mapping.get(s.charAt(right)) > 1){
                mapping.put(s.charAt(left), mapping.get(s.charAt(left))-1);
                if (mapping.get(s.charAt(left)) <= 0){
                    mapping.remove(s.charAt(left));
                }
                left ++;

            }
            longest = Math.max(longest, right-left+1);
        }
        return longest;


    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-5.py
from collections import defaultdict
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        mapping = defaultdict(int)
        longest = 0
        left = 0
        max_freq = 0
        
        for right in range(len(s)):
            mapping[s[right]] += 1
            max_freq = max(mapping.values())
            replacement = (right-left+1)-max_freq
            
            while replacement > k:
                mapping[s[left]] -= 1
                if mapping[s[left]] <= 0:
                    mapping.pop(s[left])
                left += 1
                    
                max_freq = max(mapping.values())
                replacement = (right-left+1)-max_freq
            longest = max(longest, right-left+1)
        return longest

       